[NNC] Enable unary op cpu testing

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#47374 [NNC] Enable unary op cpu testing**
* #47373 [NNC] refactor cuda half support to more general file
* #47372 [NNC] More cpu tests
* #47371 [NNC] Add more CPU Tests
* #47370 [NNC] Fix llvm min lowering for int inputs
* #47369 Add more CPU tests
* #46951 [TensorExpr][CPU] Fix bool -> int casting
* #46950 Small changes/cleanup
* #46949 Force LLVM Compilation for CPU Tests

A few small fixes needed to enable unary op cpu testing. If reviewers would prefer I split  them  up let me know.

Differential Revision: [D24805248](https://our.internmc.facebook.com/intern/diff/D24805248)

Commit:
Enable unary op cpu testing

[ghstack-poisoned]

Commit:
Update on "Enable unary op cpu testing"




[ghstack-poisoned]

Commit:
Update on "Enable unary op cpu testing"




[ghstack-poisoned]

Commit:
Update on "Enable unary op cpu testing"




[ghstack-poisoned]

Commit:
Update on "Enable unary op cpu testing"




[ghstack-poisoned]

eellison commented on torch/csrc/jit/tensorexpr/llvm_codegen.cpp:
```diff
@@ -1625,6 +1632,20 @@ void LLVMCodeGenImpl::visit(const Intrinsics* v) {
+    auto zero = llvm::ConstantInt::get(value_->getType(), 0);
+    auto neg_value = irb_.CreateSub(zero, value_);
+    auto icmp = irb_.CreateICmpSGT(value_, zero);
+    value_ = irb_.CreateSelect(icmp, value_, neg_value);
```
this could be implemented differently - any tips on how I should be going abt profiling generating assembly vs other implementations / if it matters ?

bertmaher replied:
If you want to look at the generated assembly there's a DEBUG_PRINT constant you can define somewhere (I think in llvm_codegen or llvm_jit) and that'll dump the llvm ir and the x86 assembly.  Might be worth checking out.  Could also add an operator level benchmark in benchmarks/tensorexpr

eellison replied:
I looked up with the IR and it was reasonable. I guess just whether something like a shift and & would be faster. Will look into it more as a follow up with @navahgar when we get some op benchmarks going.

eellison commented on torch/csrc/jit/tensorexpr/codegen.cpp:
```diff
@@ -48,11 +48,19 @@ std::unique_ptr<CodeGen> CreateCodeGen(
   return method(stmt, params, device, kernel_func_name);
 }
 
+ExprHandle expr_to_vec(ExprHandle v, int lanes) {
```
This is almost copied from https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/tensorexpr/llvm_codegen.cpp#L477. What's the best place to put a utility function like this, if two similar callsites warrants refactoring?

bertmaher replied:
I could see this going in any of codegen, ir, or expr.  Maybe codegen since that's where it's getting used?

Commit:
Update on "[NNC] Enable unary op cpu testing"


A few small fixes needed to enable unary op cpu testing. If reviewers would prefer I split  them  up let me know.

[ghstack-poisoned]

bertmaher commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -419,7 +419,7 @@ ExprHandle TensorExprKernel::constant(const torch::jit::Value* v) {
 ExprHandle promoteIntegerToFloat(const ExprHandle& e) {
   auto scalarType = static_cast<c10::ScalarType>(e.dtype().scalar_type());
-  if (!c10::isIntegralType(scalarType)) {
+  if (!c10::isIntegralType(scalarType, /*includeBool*/ true)) {
```
Nice catch here on bool

bertmaher commented on torch/csrc/jit/tensorexpr/llvm_jit.cpp:
```diff
@@ -101,6 +103,14 @@ class TORCH_API PytorchLLVMJITImpl {
 
+          // float -> half & half -> float conversions
+          {Mangle("__gnu_h2f_ieee"),
+           {llvm::pointerToJITTargetAddress(&c10::detail::fp16_ieee_to_fp32_value),
```
superlike

bertmaher reviewed (approved):
So excited that those tests work

Commit:
Update on "[NNC] Enable unary op cpu testing"


A few small fixes needed to enable unary op cpu testing. If reviewers would prefer I split  them  up let me know.

[ghstack-poisoned]

Commit:
Update on "[NNC] Enable unary op cpu testing"


A few small fixes needed to enable unary op cpu testing. If reviewers would prefer I split  them  up let me know.

[ghstack-poisoned]

Commit:
Update on "[NNC] Enable unary op cpu testing"


A few small fixes needed to enable unary op cpu testing. If reviewers would prefer I split  them  up let me know.

Differential Revision: [D24805248](https://our.internmc.facebook.com/intern/diff/D24805248)

[ghstack-poisoned]

Commit:
Update on "[NNC] Enable unary op cpu testing"


A few small fixes needed to enable unary op cpu testing. If reviewers would prefer I split  them  up let me know.

Differential Revision: [D24805248](https://our.internmc.facebook.com/intern/diff/D24805248)

[ghstack-poisoned]

Commit:
Update on "[NNC] Enable unary op cpu testing"


A few small fixes needed to enable unary op cpu testing. If reviewers would prefer I split  them  up let me know.

Differential Revision: [D24805248](https://our.internmc.facebook.com/intern/diff/D24805248)

[ghstack-poisoned]

Commit:
Update on "[NNC] Enable unary op cpu testing"


A few small fixes needed to enable unary op cpu testing. If reviewers would prefer I split  them  up let me know.

Differential Revision: [D24805248](https://our.internmc.facebook.com/intern/diff/D24805248)

[ghstack-poisoned]
